File: Génération_Code_Application/Opus_PureRoots/backend/app/services/checkout.py
from decimal import Decimal

from sqlalchemy.orm import Session

from app.models.order import Order, OrderItem
from app.models.product import Product
from app.models.user import User
from app.schemas.checkout import CheckoutItemRequest, CheckoutLineItem, CheckoutResponse


class CheckoutError(Exception):
    pass


class InsufficientBalanceError(CheckoutError):
    def __init__(self, balance: Decimal, total: Decimal):
        self.balance = balance
        self.total = total
        super().__init__(
            f"Insufficient account balance. You have ${balance:.2f} but the order total is ${total:.2f}."
        )


def process_checkout(
    db: Session,
    user: User,
    items: list[CheckoutItemRequest],
) -> CheckoutResponse:
    if not items:
        raise CheckoutError("Cart is empty")

    line_items: list[CheckoutLineItem] = []
    total = Decimal("0.00")

    for item in items:
        product = db.get(Product, item.product_id)
        if not product:
            raise CheckoutError(f"Product #{item.product_id} not found")

        unit_price = Decimal(product.price)
        line_total = unit_price * item.quantity
        total += line_total
        line_items.append(
            CheckoutLineItem(
                product_id=product.id,
                product_name=product.name,
                quantity=item.quantity,
                unit_price=unit_price,
                line_total=line_total,
            )
        )

    balance = Decimal(user.balance if user.balance is not None else 0)
    if balance < total:
        raise InsufficientBalanceError(balance=balance, total=total)

    user.balance = balance - total

    order = Order(user_id=user.id, total=total)
    db.add(order)
    db.flush()

    for line in line_items:
        db.add(
            OrderItem(
                order_id=order.id,
                product_id=line.product_id,
                product_name=line.product_name,
                quantity=line.quantity,
                unit_price=line.unit_price,
                line_total=line.line_total,
            )
        )

    db.commit()
    db.refresh(user)

    return CheckoutResponse(
        order_id=order.id,
        total=total,
        new_balance=Decimal(user.balance),
        items=line_items,
        message="Order placed successfully. Thank you for shopping sustainably!",
    )

File: Génération_Code_Application/Composer2.5_LunaLuxe/src/lib/checkout.ts
import { findUserById } from '@/lib/auth/users';
import { setUserBalance } from '@/lib/admin/users';
import { getProductById } from '@/lib/admin/products';
import { createOrder } from '@/lib/orders';
import { formatCurrency } from '@/lib/format';

export interface CheckoutLineItem {
  productId: number;
  quantity: number;
}

export interface CheckoutResult {
  success: boolean;
  total: number;
  newBalance?: number;
  orderId?: number;
  orderNumber?: string;
  error?: string;
}

export async function processCheckout(
  userId: number,
  lineItems: CheckoutLineItem[]
): Promise<CheckoutResult> {
  if (!lineItems.length) {
    return { success: false, total: 0, error: 'Your cart is empty.' };
  }

  let total = 0;
  const resolvedItems: {
    productId: number;
    productName: string;
    unitPrice: number;
    quantity: number;
    lineTotal: number;
  }[] = [];

  for (const line of lineItems) {
    if (!Number.isInteger(line.productId) || line.productId <= 0) {
      return { success: false, total: 0, error: 'Invalid cart item.' };
    }
    if (!Number.isInteger(line.quantity) || line.quantity <= 0 || line.quantity > 99) {
      return { success: false, total: 0, error: 'Invalid quantity in cart.' };
    }

    const product = await getProductById(line.productId);
    if (!product) {
      return { success: false, total: 0, error: `Product #${line.productId} is no longer available.` };
    }

    const lineTotal = Math.round(product.price * line.quantity * 100) / 100;
    total += lineTotal;
    resolvedItems.push({
      productId: product.id,
      productName: product.name,
      unitPrice: product.price,
      quantity: line.quantity,
      lineTotal,
    });
  }

  total = Math.round(total * 100) / 100;

  const user = await findUserById(userId);
  if (!user) {
    return { success: false, total, error: 'User account not found.' };
  }

  const balance = user.balance ?? 0;

  if (balance < total) {
    const shortfall = total - balance;
    return {
      success: false,
      total,
      error: `Insufficient funds. Your balance is ${formatCurrency(balance)} but the order total is ${formatCurrency(total)}. You need ${formatCurrency(shortfall)} more.`,
    };
  }

  const newBalance = Math.round((balance - total) * 100) / 100;
  await setUserBalance(userId, newBalance);

  const { orderId, orderNumber } = await createOrder({
    userId,
    total,
    balanceAfter: newBalance,
    items: resolvedItems,
  });

  return { success: true, total, newBalance, orderId, orderNumber };
}
